core/src/main/java/com/linecorp/armeria/server/Http2RequestDecoder.java

```diff
@@ -310,53 +318,44 @@ public int onDataRead(
         final long transferredLength = decodedReq.transferredBytes();
         if (maxContentLength > 0 && transferredLength > maxContentLength) {
             assert encoder != null;
-            final Http2Stream stream = encoder.findStream(streamId);
-            if (isWritable(stream)) {
-                final ContentTooLargeException cause =
-                        ContentTooLargeException.builder()
-                                                .maxContentLength(maxContentLength)
-                                                .contentLength(req.headers())
-                                                .transferred(transferredLength)
-                                                .build();
-
-                writeErrorResponse(streamId, req.headers(), HttpStatus.REQUEST_ENTITY_TOO_LARGE, null, cause);
-                decodedReq.abortResponse(HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause),
-                                         true);
+            final ContentTooLargeException cause =
+                    ContentTooLargeException.builder()
+                                            .maxContentLength(maxContentLength)
+                                            .contentLength(req.headers())
+                                            .transferred(transferredLength)
+                                            .build();
+
+            final HttpStatusException httpStatusException =
+                    HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause);
+            if (req.needsAggregation()) {
+                assert !req.isInitialized();
+                final StreamingDecodedHttpRequest streamingReq =
+                        decodedReq.toAbortedStreaming(inboundTrafficController, httpStatusException, true);
+                requests.put(streamId, streamingReq);
+                ctx.fireChannelRead(streamingReq);
             } else {
-                // The response has been started already. Abort the request and let the response continue.
-                decodedReq.abort();
+                decodedReq.abortResponse(httpStatusException, true);
             }
         } else if (decodedReq.isOpen()) {
             try {
                 // The decodedReq will be automatically closed if endOfStream is true.
                 decodedReq.write(HttpData.wrap(data.retain()).withEndOfStream(endOfStream));
-                if (endOfStream && decodedReq.needsAggregation()) {
+                if (endOfStream && req.needsAggregation()) {
+                    assert !req.isInitialized();
```

Comment:
Should we move the assertion into `needsAggregation()` and `isInitialized()`, given that you always do an assertion for them? e.g.

```java
boolean needsAggregation() {
  assert !needsAggregation || !initialized :
      "needsAggregation: " ... + ", isInitialized: " + ...;
  return needsAggregation;  
}
```

Reply:
There's one exception that is calling it when received an RST stream so we cannot add the assertion.
https://github.com/line/armeria/blob/86d656927dfd363f5bd371c0ace8cbe5588ac4f8/core/src/main/java/com/linecorp/armeria/server/Http2RequestDecoder.java#L396

Reply:
I see. No worries, then!